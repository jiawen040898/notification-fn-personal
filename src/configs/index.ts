export * from './aws.config';
export * from './pusher.config';
