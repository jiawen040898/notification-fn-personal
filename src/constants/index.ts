export * from './app-activity-type.enum';
export * from './assignment-due-variable-type.enum';
export * from './cache-store.enum';
export * from './email-template.enum';
export * from './event-type.enum';
export * from './icon-type.enum';
export * from './locale.enum';
