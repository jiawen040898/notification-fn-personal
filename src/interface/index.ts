export * from './company.interface';
export * from './email-request';
export * from './event-model.interface';
export * from './feedback.interface';
export * from './identity.interface';
export * from './manager-subordinate-aggregated-report-data.interface';
export * from './program.interface';
export * from './push-notification-request.interface';
export * from './ses-notification';
