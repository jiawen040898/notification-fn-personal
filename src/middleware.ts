import middy from '@middy/core';
import { initSentryMiddleware, loggerMiddleware } from '@pulsifi/fn';
import * as Sentry from '@sentry/serverless';
import { Handler } from 'aws-lambda';

import { version } from '../package.json';

initSentryMiddleware(version);

export const eventMiddleware = (handler: Handler): Handler =>
    Sentry.AWSLambda.wrapHandler(
        middy(handler).use([
            // ssm({
            //     setToContext: true,
            //     fetchData: {
            //         pusherBeamsInstanceId:
            //             '/notification-fn/PUSHER_BEAMS_INSTANCE_ID',
            //         pusherBeamsSecretKey:
            //             '/notification-fn/PUSHER_BEAMS_SECRET_KEY',
            //     },
            // }),
            loggerMiddleware(),
        ]),
        {
            timeoutWarningLimit: 5.5,
        },
        // .before(async (request) => {
        //     const data = await getInternal(
        //         ['pusherBeamsInstanceId', 'pusherBeamsSecretKey'],
        //         request,
        //     );
        //
        //     PusherConfig.beamsInstanceId = data.pusherBeamsInstanceId;
        //     PusherConfig.beamsSecretKey = data.pusherBeamsSecretKey;
        // }),
    );
