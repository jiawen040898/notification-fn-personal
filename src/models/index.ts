export * from './app-activity-message.entity';
export * from './app-message-template.entity';
export * from './email-activity-message.entity';
export * from './email-template.entity';
export * from './entities';
export * from './user-notification-setting.entity';
