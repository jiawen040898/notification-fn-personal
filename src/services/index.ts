export * from './app-notification.service';
export * from './company.service';
export * from './company-email-template.service';
export * from './email-notification.service';
export * from './email-tracking.service';
export * from './identity.service';
export * from './push-notification.service';
export * from './send-email.service';
export * from './talent-management-notification.service';
