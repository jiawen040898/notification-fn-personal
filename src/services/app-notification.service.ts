import { EventModel } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ApplicationWithdraw } from '../interface';
import { UserNotificationSetting } from '../models';
import {
    EmailNotificationService,
    IdentityService,
    PushNotificationService,
} from '.';

export class AppNotificationService {
    private readonly identityService: IdentityService;
    private readonly pushNotificationService: PushNotificationService;
    private readonly emailNotificationService: EmailNotificationService;

    constructor(private readonly dataSource: DataSource) {
        this.identityService = new IdentityService();
        this.pushNotificationService = new PushNotificationService();
        this.emailNotificationService = new EmailNotificationService(
            dataSource,
        );
    }

    getAllUserNotificationSettings = async (
        companyId: number,
        notificationEventType: string,
    ): Promise<UserNotificationSetting[]> => {
        return this.dataSource.getRepository(UserNotificationSetting).find({
            where: {
                company_id: companyId,
                notification_event_type: notificationEventType,
                is_deleted: false,
            },
        });
    };

    async notifyApplicationWithdraw(message: EventModel<JSON>): Promise<void> {
        const applicationWithdrawPayload: ApplicationWithdraw = <
            ApplicationWithdraw
        >(<unknown>message.data);

        const notificationEventType = message.event_type;

        const userNotificationSettings =
            await this.getAllUserNotificationSettings(
                applicationWithdrawPayload.company_id,
                notificationEventType,
            );

        if (userNotificationSettings.length === 0) {
            return;
        }

        const userAccountIds = userNotificationSettings.map(
            (i) => i.user_account_id,
        );
        const userAccountDetails =
            await this.identityService.getUsersByUserAccounts(userAccountIds);

        await this.emailNotificationService.sendEmailNotification(
            notificationEventType,
            userNotificationSettings,
            userAccountDetails,
        );
    }
}
