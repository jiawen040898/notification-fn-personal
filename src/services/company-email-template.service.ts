import { EventModel } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { EmailTemplateUsage } from '../constants';
import { CompanyCreated } from '../interface';
import { EmailTemplate } from '../models';

export class CompanyEmailTemplateService {
    constructor(private readonly dataSource: DataSource) {}

    async setupEmailTemplate(
        message: EventModel<CompanyCreated>,
    ): Promise<void> {
        const companyCreatedPayload = message.data;

        const defaultEmailTemplates = await this.dataSource
            .getRepository(EmailTemplate)
            .find({
                where: {
                    usage: EmailTemplateUsage.DEFAULT,
                    is_deleted: false,
                },
            });

        const companyEmailTemplates = defaultEmailTemplates.map((i) => {
            return <EmailTemplate>{
                ...i,
                id: undefined,
                company_id: companyCreatedPayload.company_id,
                usage: EmailTemplateUsage.CLIENT,
                created_by: message.user_account_id,
                created_at: undefined,
                updated_by: message.user_account_id,
                updated_at: undefined,
            };
        });

        await this.dataSource.transaction(async (manager) => {
            await manager.create(EmailTemplate, companyEmailTemplates);
        });
    }
}
