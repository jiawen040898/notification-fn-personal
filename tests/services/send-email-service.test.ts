import { DataSource } from 'typeorm';

import {
    DefaultLocale,
    EmailTemplateCommunicationType,
} from '../../src/constants';
import { SendEmailService } from '../../src/services';
import { testData as generalTestData } from '../fixtures';
import { getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/send-email-service.test-data';

jest.mock('@pulsifi/fn/utils/logger.util');

jest.mock('../../src/services/company.service', () => {
    const mBInstance = {
        getCompanyData: jest.fn(() => [generalTestData.companyDataResponse]),
        getCompanyLocaleByCompanyId: jest.fn(() => [DefaultLocale]),
    };
    const mB = jest.fn(() => mBInstance);
    return { CompanyService: mB };
});

describe('getEmailTemplate', () => {
    let dataSource: DataSource;
    let sut: SendEmailService;

    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesToBeAddedToDb,
        );

        sut = new SendEmailService(dataSource);
    });

    afterAll(() => {
        dataSource.destroy();
    });

    it('should pass getting email template if email communication type is not provided', async () => {
        // Act
        const actual = await sut.getEmailTemplate(
            testData.emailTemplate1.company_id,
            testData.emailTemplate1.id,
        );

        // Assert
        expect(actual).toMatchSnapshot({
            created_at: expect.any(Date),
            updated_at: expect.any(Date),
        });
    });

    it('should pass getting email template if email template id is not provided', async () => {
        // Act
        const actual = await sut.getEmailTemplate(
            testData.emailTemplate1.company_id,
            undefined,
            EmailTemplateCommunicationType.APPLICATION_SHORTLIST,
        );

        // Assert
        expect(actual).toMatchSnapshot({
            created_at: expect.any(Date),
            updated_at: expect.any(Date),
        });
    });

    it('should pass getting correct email template when there are two and above same email communication type in a company', async () => {
        // Act
        const actual = await sut.getEmailTemplate(
            testData.emailTemplate2.company_id,
            testData.emailTemplate2.id,
            EmailTemplateCommunicationType.APPLICATION_SHORTLIST,
        );

        // Assert
        expect(actual).toMatchSnapshot({
            created_at: expect.any(Date),
            updated_at: expect.any(Date),
        });
    });

    it('should fail getting email template', async () => {
        // Act
        const action = () =>
            sut.getEmailTemplate(testData.emailTemplate1.company_id);

        // Assert
        await expect(action()).rejects.toThrow('No email template found');
    });
});
