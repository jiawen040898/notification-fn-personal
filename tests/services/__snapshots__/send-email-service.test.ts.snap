// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEmailTemplate should pass getting correct email template when there are two and above same email communication type in a company 1`] = `
{
  "body_in_html": "<html>Hello {{first_name}} {{year}}</html>",
  "company_id": 1,
  "created_at": Any<Date>,
  "created_by": 1,
  "email_communication_type": "application_shortlist",
  "id": 2,
  "is_deleted": false,
  "language": "en-US",
  "notification_event_type": null,
  "sender_name": "Sender",
  "subject": "Hello world {{job_title}}",
  "updated_at": Any<Date>,
  "updated_by": 1,
  "usage": "client",
}
`;

exports[`getEmailTemplate should pass getting email template if email communication type is not provided 1`] = `
{
  "body_in_html": "<html>Hello {{first_name}} {{year}}</html>",
  "company_id": 1,
  "created_at": Any<Date>,
  "created_by": 1,
  "email_communication_type": "application_shortlist",
  "id": 1,
  "is_deleted": false,
  "language": "en-US",
  "notification_event_type": null,
  "sender_name": "Sender",
  "subject": "Hello world {{job_title}}",
  "updated_at": Any<Date>,
  "updated_by": 1,
  "usage": "client",
}
`;

exports[`getEmailTemplate should pass getting email template if email template id is not provided 1`] = `
{
  "body_in_html": "<html>Hello {{first_name}} {{year}}</html>",
  "company_id": 1,
  "created_at": Any<Date>,
  "created_by": 1,
  "email_communication_type": "application_shortlist",
  "id": 1,
  "is_deleted": false,
  "language": "en-US",
  "notification_event_type": null,
  "sender_name": "Sender",
  "subject": "Hello world {{job_title}}",
  "updated_at": Any<Date>,
  "updated_by": 1,
  "usage": "client",
}
`;
